File: geometry_generators/generateSpheresResolved.m
function sphere_data_resolved = generateSpheresResolved(num_target_spheres, min_r, max_r, outer_radius, min_separation_distance)
% Generates a set of non-intersecting polydisperse spheres confined inside a
% spherical container of radius outer_radius (centered at the origin).
%
% Inputs:
%   num_target_spheres      - The desired number of spheres to generate.
%   min_r                   - Minimum radius for an interior sphere.
%   max_r                   - Maximum radius for an interior sphere.
%   outer_radius            - Radius of the spherical container.
%   min_separation_distance - Minimum gap between sphere surfaces (and between
%                             a sphere surface and the container wall).
%
% Output:
%   sphere_data_resolved - An N x 4 matrix where N is the number of successfully placed spheres.
%                          Each row is [x, y, z, r] representing center coordinates and radius.

    sphere_data_resolved = []; % Initialize empty matrix for spheres
    max_attempts_per_sphere = 5000; % Max attempts to place a single sphere

    fprintf('Attempting to generate %d non-intersecting spheres in a sphere of radius %.4f...\n', ...
            num_target_spheres, outer_radius);

    for i = 1:num_target_spheres
        found_position = false;
        current_sphere_attempts = 0;

        while ~found_position && current_sphere_attempts < max_attempts_per_sphere
            % 1. Generate a random radius
            r_new = min_r + (max_r - min_r) * rand();

            % 2. Generate a random center uniformly inside the ball of radius
            %    (outer_radius - r_new - min_separation_distance) so the new
            %    sphere stays fully inside the container, clear of the wall.
            R_place = outer_radius - r_new - min_separation_distance;
            if R_place <= 0
                current_sphere_attempts = current_sphere_attempts + 1;
                continue; % sphere too big to fit in the container
            end
            % Uniform-in-ball sampling: random direction * radius scaled by rand^(1/3)
            dir = randn(1, 3);
            dir = dir / norm(dir);
            radial = R_place * rand()^(1/3);
            center_new = radial * dir;

            intersects = false;

            % 3. Check for intersection with all previously placed spheres
            for j = 1:size(sphere_data_resolved, 1)
                center_existing = sphere_data_resolved(j, 1:3);
                r_existing = sphere_data_resolved(j, 4);

                % Plain Euclidean distance between centers (no periodicity)
                dist_centers = norm(center_new - center_existing);

                % Sum of radii plus required gap
                sum_radii = r_new + r_existing + min_separation_distance;

                % Check for intersection
                if dist_centers < sum_radii
                    intersects = true;
                    break; % Intersection found, break from inner loop and try a new random sphere
                end
            end

            % 4. If no intersection, add the new sphere to the list
            if ~intersects
                sphere_data_resolved = [sphere_data_resolved; center_new, r_new];
                found_position = true;
            end
            current_sphere_attempts = current_sphere_attempts + 1;
        end

        if ~found_position
            fprintf('Warning: Could not place sphere %d after %d attempts.\n', i, max_attempts_per_sphere);
            fprintf('         Generation stopped at %d spheres. This usually means the packing limit is reached.\n', size(sphere_data_resolved, 1));
            break; % Stop trying to add more spheres if one cannot be placed
        end
    end
    fprintf('Successfully generated %d non-intersecting spheres.\n', size(sphere_data_resolved, 1));
end

File: geometry_generators/generate_spheres_resolved.py
"""Generate non-intersecting polydisperse spheres inside a spherical container.

Python port of generateSpheresResolved.m / spheres.m (without plotting). Spheres are
confined inside a ball of radius ``outer_radius`` centered at the origin, kept clear of
each other and of the container wall by at least ``min_separation_distance``.

Run as a script to reproduce the demo (~100 spheres in a unit ball) and write the
result as a 4-column ``[x y z r]`` text file into this folder, byte-compatible with the
MATLAB output of spheres.m::

    python geometry_generators/generate_spheres_resolved.py
"""

import os

import numpy as np

MAX_ATTEMPTS_PER_SPHERE = 5000  # Max attempts to place a single sphere


def generate_spheres_resolved(num_target_spheres, min_r, max_r, outer_radius,
                              min_separation_distance, rng=None):
    """Place non-intersecting spheres inside a spherical container.

    Parameters
    ----------
    num_target_spheres : int
        Desired number of interior spheres to generate.
    min_r, max_r : float
        Minimum / maximum radius of an interior sphere.
    outer_radius : float
        Radius of the spherical container (centered at the origin).
    min_separation_distance : float
        Minimum gap between sphere surfaces (and between a sphere surface and the wall).
    rng : numpy.random.Generator, optional
        Random generator (for reproducibility). Defaults to ``np.random.default_rng()``.

    Returns
    -------
    numpy.ndarray
        ``(N, 4)`` array; each row is ``[x, y, z, r]`` for a placed sphere.
    """
    if rng is None:
        rng = np.random.default_rng()

    spheres = []  # list of [x, y, z, r]

    print(f"Attempting to generate {num_target_spheres} non-intersecting spheres "
          f"in a sphere of radius {outer_radius:.4f}...")

    for i in range(num_target_spheres):
        found_position = False
        attempts = 0

        while not found_position and attempts < MAX_ATTEMPTS_PER_SPHERE:
            attempts += 1

            # 1. Random radius
            r_new = min_r + (max_r - min_r) * rng.random()

            # 2. Random center uniformly inside the ball of radius
            #    (outer_radius - r_new - min_separation_distance) so the new sphere
            #    stays fully inside the container, clear of the wall.
            R_place = outer_radius - r_new - min_separation_distance
            if R_place <= 0:
                continue  # sphere too big to fit in the container

            # Uniform-in-ball: random direction * radius scaled by rand^(1/3)
            direction = rng.standard_normal(3)
            direction /= np.linalg.norm(direction)
            radial = R_place * rng.random() ** (1.0 / 3.0)
            center_new = radial * direction

            # 3. Check for intersection with previously placed spheres (Euclidean)
            intersects = False
            for x, y, z, r_existing in spheres:
                dist = np.linalg.norm(center_new - np.array([x, y, z]))
                if dist < r_new + r_existing + min_separation_distance:
                    intersects = True
                    break

            # 4. Accept if clear
            if not intersects:
                spheres.append([center_new[0], center_new[1], center_new[2], r_new])
                found_position = True

        if not found_position:
            print(f"Warning: Could not place sphere {i + 1} after "
                  f"{MAX_ATTEMPTS_PER_SPHERE} attempts.")
            print(f"         Generation stopped at {len(spheres)} spheres. "
                  f"This usually means the packing limit is reached.")
            break

    print(f"Successfully generated {len(spheres)} non-intersecting spheres.")
    return np.array(spheres, dtype=float).reshape(-1, 4)


if __name__ == "__main__":
    # --- Container + packing parameters (mirrors spheres.m) ---
    outer_radius = 1.0              # radius of the spherical container
    Nptcl = 10                     # number of interior spheres to attempt
    min_r = 0.1                    # minimum interior-sphere radius (~0.1)
    max_r = 0.2                    # maximum interior-sphere radius (~0.1)
    min_separation_distance = 0.001  # minimum surface-to-surface gap

    rng = np.random.default_rng(1)  # seeded for reproducibility

    sphere_data = generate_spheres_resolved(
        Nptcl, min_r, max_r, outer_radius, min_separation_distance, rng=rng)

    # --- Report total volume + volume fraction ---
    tot_vol = float(np.sum(4.0 * np.pi * sphere_data[:, 3] ** 3 / 3.0))
    container_vol = 4.0 * np.pi * outer_radius ** 3 / 3.0
    print(f"\n Total interior-sphere volume is {tot_vol:f}.")
    print(f" Volume fraction (relative to container) is {tot_vol / container_vol:f}.")

    # --- Write to file in this folder, matching the MATLAB %.8f format ---
    here = os.path.dirname(os.path.abspath(__file__))
    filename = os.path.join(here, f"sphere_data_{sphere_data.shape[0]}_ball.txt")
    np.savetxt(filename, sphere_data, fmt="%.8f",
               header=f"outer_radius = {outer_radius:.8f}")
    print(f" Wrote {sphere_data.shape[0]} spheres to {filename}")
